Explain interaction of caching, status and extensions (#826)
moq-wg/moq-transport: draft-ietf-moq-transport.md

Relays are endpoints, which means they terminate Transport Sessions in order to
have visibility of MoQ Object metadata.

## Caching Relays

Relays MAY cache Objects, but are not required to.

A caching relay saves Objects to its cache identified by the Object's
Full Track Name, Group ID and Object ID.
If multiple objects are received with the same Full Track Name,
Group ID and Object ID, Relays MAY ignore subsequently received Objects
or MAY use them to update the cache. Implementations that update the
cache need to protect against cache poisoning.  The only fields that can be
updated are the following:

1. Object Status can transition from any status to Object Not Exists in cases
   where the object is no longer available
2. Object Header Extensions can be added, removed or updated

> An endpoint that receives a duplicate object with fields that differ from a

Reviewer: If I understand correctly, we're really saying that the content of an Object can never change, correct?  If so, it might be simpler to say that.

Author: The status can also not change, say from Normal to End of Group.  Technically the subgroup and priority can't change either per this language.  I'll add it all explicitly.

Relays are endpoints, which means they terminate Transport Sessions in order to
have visibility of MoQ Object metadata.

## Caching Relays

Relays MAY cache Objects, but are not required to.

A caching relay saves Objects to its cache identified by the Object's

> Full Track Name, Group ID and Object ID.

Reviewer 2: To support multi-publisher, we need publisher ID as well.

Reviewer 2: Correct, one or more publishers index their group/object IDs independently without any global synchronization. This can result in publisher 1 and 2 overlapping or appearing out of order, unless we distinguish that there are two publishers with their own group/object IDs.

Author: If you want to live on the edge with multiple publishers publishing to the same track in an uncoordinated fashion, you still need to abide but the referenced MUST.  The best option is to embed your "Publisher ID" in the track namespace or name.  I imagine with forthcoming changes addressing #615 will make that easier.

If multiple objects are received with the same Full Track Name,
Group ID and Object ID, Relays MAY ignore subsequently received Objects
or MAY use them to update the cache. Implementations that update the
cache need to protect against cache poisoning.  The only fields that can be
updated are the following:

1. Object Status can transition from any status to Object Not Exists in cases
   where the object is no longer available

> 2. Object Header Extensions can be added, removed or updated

Reviewer 2: Headers, specifically extension headers, are problem as some extension headers may be specific relay by relay (e.g., Relay 5 added headers that relay 14 never sees). Caching is a bit tricky in this sense where mutable headers are not the same relay by relay. In a distributed deep cache, the headers may not exist at all due to which relay populated the deep cache. When pulling from a deep cache (distributed, not directly on the relay), the fetch may not match what would have been there had the relay originally received the data based on subscribe. Are relays to update the extension headers upon fetch response?

> 2. Object Header Extensions can be added, removed or updated

Reviewer: ```suggestion
2. Object Header Extensions can be added, removed or updated, subject
to the constraints of the specific header extension.
```

An endpoint that receives a duplicate object with an illegal Object Status

> change, or a Forwarding Preference, Subgroup ID, Priority or Payload that

Reviewer: Are we sure we can guarantee that an illegal Object Status change can't happen in a distributed network?  Also, I'm not sure it's that valuable to make the transitions illegal?  It seems the most important bit is what I can cache. (ie: negative caching)

Reviewer: Fair point.  I guess in that case the "Normal" Object should be ignored under the assumption that the "Does not exist" status is more recent?

Author: So some transitions will always be illegal:  eg: Normal, End of Group, End of Track can never change amongst each other.  But I think it's fine to say if you get one after receiving Does Not Exist, you can assume Does Not Exist is more recent and ignore the new status?

Author: I added more text explaining this, and relaxed from "session error" to Malformed track.  Please take a look.

> differ from a previous version MUST close the session with a Protocol

Reviewer 2: At some point cache invalidation (aka purge) will be needed.  We'll need to be able to from a publisher indicate, maybe in status, that object(s) should be purged.  

In addition, time to live needs to be taken into account. What is the retention policy and how will a subscriber/fetcher know that cache will never have requested data because they are asking for data outside of the retention policy?

Relays are endpoints, which means they terminate Transport Sessions in order to
have visibility of MoQ Object metadata.

## Caching Relays

Relays MAY cache Objects, but are not required to.

A caching relay saves Objects to its cache identified by the Object's
Full Track Name, Group ID and Object ID.
If multiple objects are received with the same Full Track Name,
Group ID and Object ID, Relays MAY ignore subsequently received Objects
or MAY use them to update the cache. Implementations that update the

> cache need to protect against cache poisoning.  The only fields that can be

Reviewer 3: ```suggestion
cache need to protect against cache poisoning.  The
same object SHOULD NOT be forwarded more than once on the same subscription.
The only fields that can be
```

updated are the following:

1. Object Status can transition from any status to Object Not Exists in cases
   where the object is no longer available.  Transitions between Normal, End
   of Group and End of Track are invalid.
3. Object Header Extensions can be added, removed or updated, subject
   to the constraints of the specific header extension.

An endpoint that receives a duplicate Object with an invalid Object Status
change, or a Forwarding Preference, Subgroup ID, Priority or Payload that

> differ from a previous version MUST treat the track as Malformed.

Reviewer 4: This seems to contradict the earlier statement that relays MAY update the cache.

Author: There are only specific fields in a cache that can be updated without breaking the rules.  Maybe change 'MAY update the cache' to MAY update allowed fields in the cache'?

track within the subscription range is forwarded to each active
subscriber, dependent on the congestion response.

> Relays MUST be able to process objects for the same Full Track Name from

Reviewer 2: publisher ID is needed to support multiple publishers to the same namespace tuple/name.

Reviewer 3: I will not be in favor adding any text that enforces a namespace design/structure within the core transport. An application can choose to have publisherId as a tuple or use one track to have multiple publisher or have publisher come acoss multiple connections (says make before break).

multiple publishers and forward objects to active matching subscriptions.  The

> same object SHOULD NOT be forwarded more than once to the same subscriber.

Reviewer: I think we want to say for the same subscription?  Because a relay might get an Object once via SUBSCRIBE and later via FETCH?

track within the subscription range is forwarded to each active
subscriber, dependent on the congestion response.

Relays MUST be able to process objects for the same Full Track Name from
multiple publishers and forward objects to active matching subscriptions.  The

> same object SHOULD NOT be forwarded more than once to the same subscriber.

Author: ```suggestion
same object SHOULD NOT be forwarded more than once on the same subscription.
```

track within the subscription range is forwarded to each active
subscriber, dependent on the congestion response.

Relays MUST be able to process objects for the same Full Track Name from
multiple publishers and forward objects to active matching subscriptions.  The

> same object SHOULD NOT be forwarded more than once on the same subscription.

Reviewer 3: I would remove the last sentence and move it up.
